roaringbitmap/src/main/java/org/roaringbitmap/OrderedWriter.java

```diff
@@ -0,0 +1,88 @@
+package org.roaringbitmap;
+
+import java.util.Arrays;
+
+public class OrderedWriter {
+
+  private static final int WORD_COUNT = 1 << 10;
+  private final long[] bitmap;
+  private final RoaringBitmap underlying;
+
+  private short currentKey;
+  private boolean dirty = false;
+
+  public OrderedWriter(RoaringBitmap underlying) {
+    this.underlying = underlying;
+    this.bitmap = new long[WORD_COUNT];
+  }
+
+  /**
+   * Adds the value to the underlying bitmap.
+   * @param value the value to add.
+   */
+  public void add(int value) {
+    short key = Util.highbits(value);
+    short low = Util.lowbits(value);
+    if (Util.compareUnsigned(key, currentKey) < 0) {
+      throw new IllegalStateException("Must write in ascending key order");
+    }
+    if (key != currentKey) {
+      flush();
+    }
+    int ulow = low & 0xFFFF;
+    bitmap[(ulow >>> 6)] |= (1L << ulow);
+    currentKey = key;
+    dirty = true;
+  }
+
+  /**
+   * Ensures that any buffered additions are flushed to the underlying bitmap.
+   */
+  public void flush() {
+    if (dirty) {
+      int cardinality = computeCardinality();
+      underlying.highLowContainer.append(currentKey, chooseBestContainer(cardinality));
+      clearBitmap();
+      dirty = false;
+    }
+  }
+
+  private Container chooseBestContainer(int cardinality) {
```

Comment:
You can simplify code as follows:

Container c =  new BitmapContainer(bitmap,-1).repairAfterLazy().runOptimize();
if(c instanceof BitmapContainer) {
  return c.clone();
} else {
  return c;
}

Main benefit is to reduce code redundancy.

(Your code is fine and I'm happy to merge it as-is, but please consider my proposal.)

Reply:
I think that's a very helpful suggestion. I made this change.